storage/pruner/pruner.go

```diff
@@ -0,0 +1,374 @@
+package pruner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+	pebbleStorage "github.com/onflow/flow-go/storage/pebble"
+)
+
+type PruneMethod int
+
+const (
+	// PruneByHeight prunes data using finalized block heights
+	// This is should be used when a db does not include a view index
+	PruneByHeight PruneMethod = iota + 1
+
+	// PruneByView prunes data using views
+	// This is the simpler and more efficient method, but requires the node to have a view index
+	PruneByView
+
+	// DefaultPruneInterval is the default interval between pruning cycles
+	DefaultPruneInterval = 5 * time.Minute
+
+	// DefaultThrottleDelay is the default delay between pruning individual blocks
+	DefaultThrottleDelay = 20 * time.Millisecond
+)
+
+type Pruner struct {
+	component.Component
+
+	log zerolog.Logger
+	db  *pebble.DB
+
+	headers        storage.Headers
+	chunkDataPacks storage.ChunkDataPacks
+
+	targetKeepBlocks uint64 // number of blocks to keep below the highest block
+	pruneThreshold   uint64 // number of blocks to exceed the target before pruning
+	pruneMethod      PruneMethod
+
+	lowestHeader     *flow.Header
+	rootSealedHeader *flow.Header
+	progress         storage.ConsumerProgress
+
+	producers []*BlockProcessedProducer
+	mu        sync.RWMutex
+}
+
+func New(
+	log zerolog.Logger,
+	db *pebble.DB,
+	headers storage.Headers,
+	chunkDataPacks storage.ChunkDataPacks,
+	rootSealedHeader *flow.Header,
+	progress storage.ConsumerProgress,
+	targetKeepBlocks uint64,
+	pruneThreshold uint64,
+) *Pruner {
+	pruneMethod := PruneByHeight
+	if progress.Consumer() == module.ConsumeProgressProtocolDBPrunerByView {
+		pruneMethod = PruneByView
+	}
+
+	p := &Pruner{
+		log: log.With().Str("component", "protocoldb_pruner").Logger(),
+		db:  db,
+
+		headers:          headers,
+		chunkDataPacks:   chunkDataPacks,
+		targetKeepBlocks: targetKeepBlocks,
+		pruneThreshold:   pruneThreshold,
+		rootSealedHeader: rootSealedHeader,
+
+		progress:    progress,
+		pruneMethod: pruneMethod,
+	}
+
+	p.Component = component.NewComponentManagerBuilder().
+		AddWorker(p.loop).
```

Comment:
I usually prefer to make a core engine that being wrapped by the component. 

The component wrapper will trigger the timer, whereas the core is a passive module, that provides a `Prune` method, if core has not finished the previous prune, then the call will be ignored. 

This would make it easy for writing tests for the pruning logic in core. You can refer to the ingestion core engine, where as the ingestion machine is the wrapper.

We probably don't need much tests for the wrapper, as it's just handling the startup and a timer.